src/ipvs/ip_vs_proto_tcp.c

```diff
@@ -720,6 +764,14 @@ static int tcp_fnat_in_handler(struct dp_vs_proto *proto,
     if (th->syn && !th->ack) {
         tcp_in_remove_ts(th);
         tcp_in_init_seq(conn, mbuf, th);
+        if (tcp_in_check_toa(conn, th, &dcdn_addr) == EDPVS_OK) {
+            conn->dcdn_found = true;
+            conn->dcdn_addr = dcdn_addr;
```

Comment:
It seems every TCP packet go through the `tcp_in_check_toa`. It may be optimized with the `dcdn_found` flag by checking the toa only when the flag is not set.